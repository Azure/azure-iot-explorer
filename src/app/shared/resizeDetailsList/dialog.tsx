/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { DefaultButton, Dialog, DialogFooter, DialogType, IColumn, IDetailsList, ITextField, PrimaryButton, TextField } from '@fluentui/react';
import { ResourceKeys } from '../../../localization/resourceKeys';

interface ResizeDetailsListDialog {
    columnToEdit: React.MutableRefObject<IColumn>;
    detailsListRef: React.MutableRefObject<IDetailsList>;
    isDialogHidden: boolean;
    setIsDialogHidden: (isDialogHidden: boolean) => void;
}

export const ResizeDetailsListDialog: React.FC<ResizeDetailsListDialog> = ({columnToEdit, detailsListRef, isDialogHidden, setIsDialogHidden}) => {
    const { t } = useTranslation();
    const textfieldRef = React.useRef<ITextField>(null);
    const input = React.useRef<number | null>(null);

    const resizeDialogContentProps = {
        closeButtonAriaLabel: 'Close',
        subText: t(ResourceKeys.resizableDetailsList.content.subText),
        title: t(ResourceKeys.resizableDetailsList.content.title),
        type: DialogType.normal
    };

    const modalProps = {
        isBlocking: false,
        styles: { main: { maxWidth: 450 } },
        titleAriaId: 'Dialog'
    };

    const hideDialog = () => setIsDialogHidden(true);

    const confirmDialog = () => {
        const detailsList = detailsListRef.current;

        if (textfieldRef.current) {
            input.current = Number(textfieldRef.current.value);
        }
        if (columnToEdit.current && input.current && detailsList) {
            const width = input.current;
            detailsList.updateColumn(columnToEdit.current, { width });
        }

        input.current = null;
        hideDialog();
    };

    return (
        <Dialog
            hidden={isDialogHidden}
            onDismiss={hideDialog}
            dialogContentProps={resizeDialogContentProps}
            modalProps={modalProps}
        >
            <TextField
                componentRef={textfieldRef}
                ariaLabel={t(ResourceKeys.resizableDetailsList.content.subText)}
            />
            <DialogFooter>
                <PrimaryButton onClick={confirmDialog} text={t(ResourceKeys.resizableDetailsList.buttons.resize)} />
                <DefaultButton onClick={hideDialog} text={t(ResourceKeys.resizableDetailsList.buttons.cancel)} />
            </DialogFooter>
        </Dialog>
    );
};
