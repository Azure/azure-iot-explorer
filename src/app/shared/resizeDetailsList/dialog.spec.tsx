/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { shallow } from 'enzyme';
import { PrimaryButton } from '@fluentui/react';
import { ResizeDetailsListDialog } from './dialog';

describe('ResizeDetailsListDialog', () => {
    it('matches snapshot', () => {
        expect(shallow(<ResizeDetailsListDialog
                columnToEdit={undefined as any}
                detailsListRef={undefined as any}
                isDialogHidden={true}
                setIsDialogHidden={jest.fn()}
        />)).toMatchSnapshot();
    });

    it('matches snapshot when dialog is shown', () => {
        expect(shallow(<ResizeDetailsListDialog
                columnToEdit={undefined as any}
                detailsListRef={undefined as any}
                isDialogHidden={false}
                setIsDialogHidden={jest.fn()}
        />)).toMatchSnapshot();
    });

    it('clicks confirm button calls the action with expected parameter', () => {
        const setIsDialogHidden = jest.fn();
        const wrapper = shallow(<ResizeDetailsListDialog
            columnToEdit={{current: jest.fn()} as any}
            detailsListRef={{current: jest.fn()} as any}
            isDialogHidden={false}
            setIsDialogHidden={setIsDialogHidden}
        />);

        wrapper.find(PrimaryButton).props().onClick(undefined);
        expect(setIsDialogHidden).toHaveBeenCalledWith(true);
    });
});
