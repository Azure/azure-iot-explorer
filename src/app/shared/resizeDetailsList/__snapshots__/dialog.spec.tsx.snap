// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResizeDetailsListDialog matches snapshot 1`] = `
<Dialog
  dialogContentProps={
    Object {
      "closeButtonAriaLabel": "Close",
      "subText": "resizableDetailsList.content.subText",
      "title": "resizableDetailsList.content.title",
      "type": 0,
    }
  }
  hidden={true}
  modalProps={
    Object {
      "isBlocking": false,
      "styles": Object {
        "main": Object {
          "maxWidth": 450,
        },
      },
      "titleAriaId": "Dialog",
    }
  }
  onDismiss={[Function]}
>
  <StyledTextFieldBase
    ariaLabel="resizableDetailsList.content.subText"
    componentRef={
      Object {
        "current": null,
      }
    }
  />
  <StyledDialogFooterBase>
    <CustomizedPrimaryButton
      onClick={[Function]}
      text="resizableDetailsList.buttons.resize"
    />
    <CustomizedDefaultButton
      onClick={[Function]}
      text="resizableDetailsList.buttons.cancel"
    />
  </StyledDialogFooterBase>
</Dialog>
`;

exports[`ResizeDetailsListDialog matches snapshot when dialog is shown 1`] = `
<Dialog
  dialogContentProps={
    Object {
      "closeButtonAriaLabel": "Close",
      "subText": "resizableDetailsList.content.subText",
      "title": "resizableDetailsList.content.title",
      "type": 0,
    }
  }
  hidden={false}
  modalProps={
    Object {
      "isBlocking": false,
      "styles": Object {
        "main": Object {
          "maxWidth": 450,
        },
      },
      "titleAriaId": "Dialog",
    }
  }
  onDismiss={[Function]}
>
  <StyledTextFieldBase
    ariaLabel="resizableDetailsList.content.subText"
    componentRef={
      Object {
        "current": null,
      }
    }
  />
  <StyledDialogFooterBase>
    <CustomizedPrimaryButton
      onClick={[Function]}
      text="resizableDetailsList.buttons.resize"
    />
    <CustomizedDefaultButton
      onClick={[Function]}
      text="resizableDetailsList.buttons.cancel"
    />
  </StyledDialogFooterBase>
</Dialog>
`;
